tools/tpm2_duplicate.c

```diff
@@ -0,0 +1,423 @@
+                             NULL, 0);
+
+    return *opts != NULL;
+}
+
+/**
+ * Check all options and report as many errors as possible via LOG_ERR.
+ * @return
+ *  true on success, false on failure.
+ */
+static bool check_options(void) {
+
+    bool result = true;
+
+    /* Check for NULL alg & (keyin | keyout) */
+    if (ctx.flags.g == 0) {
+        LOG_ERR("Expected key type to be specified via \"-G\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.key_type != TPM2_ALG_NULL) {
+        if((ctx.flags.i == 0) && (ctx.flags.o == 0)) {
+            LOG_ERR("Expected in or out encryption key file \"-k/K\","
+                    " missing option.");
+            result = false;
+        }
+        if (ctx.flags.i && ctx.flags.o) {
+            LOG_ERR("Expected either in or out encryption key file \"-k/K\","
+                    " conflicting options.");
+            result = false;
+        }
+    }
+    else
+    {
+        if (ctx.flags.i || ctx.flags.o) {
+            LOG_ERR("Expected neither in nor out encryption key file \"-k/K\","
+                    " conflicting options.");
+            result = false;
+        }
+    }
+
+    if (ctx.flags.C == 0) {
+        LOG_ERR("Expected new parent object to be specified via \"-C\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.c == 0) {
+        LOG_ERR("Expected object to be specified via \"-c\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.s == 0) {
+        LOG_ERR("Expected encrypted seed out filename to be specified via \"-S\","
+                " missing option.");
+        result = false;
+    }
+
+    if (ctx.flags.r == 0) {
+        LOG_ERR("Expected private key out filename to be specified via \"-r\","
+                " missing option.");
+        result = false;
+    }
+
+    return result;
+}
+
+static bool load_object(ESYS_CONTEXT *ectx, const char *arg, tpm2_loaded_object * obj) {
+    bool result = true;
+    tpm2_object_load_rc rc = tpm2_util_object_load(ectx, arg, obj);
+
+    if (rc != olrc_error) {
+        if (!obj->tr_handle) {
+            result = tpm2_util_sys_handle_to_esys_handle(ectx, obj->handle, &obj->tr_handle);
+        }
+    }
+    return result;
+}
+
+static bool set_key_algorithm(TPMI_ALG_PUBLIC alg, TPMT_SYM_DEF_OBJECT * obj) {
+    bool result = true;
+    switch (alg) {
+
+    case TPM2_ALG_AES :
+        obj->algorithm = TPM2_ALG_AES;
+        obj->keyBits.aes = 128;
+        obj->mode.aes = TPM2_ALG_CFB;
+        break;
+
+    case TPM2_ALG_NULL :
+        obj->algorithm = TPM2_ALG_NULL;
+        break;
+
+    default:
+        LOG_ERR("The algorithm type input(%4.4x) is not supported!", alg);
+        result = false;
+        break;
+    }
+
+    return result;
+}
+
+static bool load_sym_key(TPM2B_DATA * in_key) {
+    FILE *f;
+    unsigned long file_size = 0;
+    bool result;
```

Comment:
We scope variables as close to use as we can. With the way your labels work, you can do FILE *f = fopen();

Reply:
Function removed